src/lem/circuit.rs

```diff
@@ -58,268 +57,369 @@ use super::{
 
 /// Manages global allocations for constants in a constraint system
 #[derive(Default)]
-pub(crate) struct GlobalAllocator<F: LurkField>(HashMap<FWrap<F>, AllocatedNum<F>>);
+pub struct GlobalAllocator<F: LurkField>(HashMap<FWrap<F>, AllocatedNum<F>>);
 
 #[inline]
 fn allocate_num<F: LurkField, CS: ConstraintSystem<F>>(
     cs: &mut CS,
     namespace: &str,
     value: F,
-) -> Result<AllocatedNum<F>> {
+) -> Result<AllocatedNum<F>, SynthesisError> {
     AllocatedNum::alloc(cs.namespace(|| namespace), || Ok(value))
-        .with_context(|| format!("allocation for '{namespace}' failed"))
+}
+
+impl<F: LurkField> GlobalAllocator<F> {
+    /// Checks if the allocation for a numeric variable has already been cached.
+    /// If so, don't do anything. Otherwise, allocate and cache it.
+    fn alloc_const<CS: ConstraintSystem<F>>(&mut self, cs: &mut CS, f: F) {
+        self.0.entry(FWrap(f)).or_insert_with(|| {
+            allocate_constant(
+                &mut cs.namespace(|| format!("allocate constant {}", f.hex_digits())),
+                f,
+            )
+        });
+    }
+
+    #[inline]
+    fn get_allocated_const(&self, f: F) -> Result<&AllocatedNum<F>> {
+        self.0
+            .get(&FWrap(f))
+            .ok_or_else(|| anyhow!("Global allocation not found for {}", f.hex_digits()))
+    }
+
+    #[inline]
+    fn get_allocated_const_cloned(&self, f: F) -> Result<AllocatedNum<F>> {
+        self.get_allocated_const(f).cloned()
+    }
```

Comment:
Nit: two suggestions, [C-GETTER](https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter), and [C-CALLER_CONTROL](https://rust-lang.github.io/api-guidelines/flexibility.html#caller-decides-where-to-copy-and-place-data-c-caller-control) which hints that `get_allocated_const` would perhaps suffice.

Reply:
Without the `cloned` version, I would have to manually write `clone()` outside many many times. It's better to have this shortcut to avoid repeating myself